src/controllers/legacycontrollermappingfilehandler.cpp

```diff
@@ -100,6 +101,85 @@ void LegacyControllerMappingFileHandler::parseMappingInfo(
     mapping->setWikiLink(wiki.isNull() ? "" : wiki.text());
 }
 
+void LegacyControllerMappingFileHandler::parseMappingSettings(
+        const QDomElement& root, std::shared_ptr<LegacyControllerMapping> mapping) const {
+    if (root.isNull() || !mapping) {
+        return;
+    }
+
+    QDomElement settings = root.firstChildElement("settings");
+    if (settings.isNull()) {
+        return;
+    }
+
+    std::unique_ptr<LegacyControllerSettingsLayoutContainer> settingLayout =
+            std::make_unique<LegacyControllerSettingsLayoutContainer>(
+                    LegacyControllerSettingsLayoutContainer::Disposition::
+                            VERTICAL);
+    parseMappingSettingsElement(settings, mapping, settingLayout);
+    mapping->setSettingLayout(std::move(settingLayout));
+}
+
+void LegacyControllerMappingFileHandler::parseMappingSettingsElement(
+        const QDomElement& current,
+        std::shared_ptr<LegacyControllerMapping> pMapping,
+        const std::unique_ptr<LegacyControllerSettingsLayoutContainer>& layout)
+        const {
+    for (QDomElement element = current.firstChildElement();
+            !element.isNull();
+            element = element.nextSiblingElement()) {
+        const QString& tagName = element.tagName().toLower();
+        if (tagName == "option") {
+            std::shared_ptr<AbstractLegacyControllerSetting> pSetting(
+                    LegacyControllerSettingBuilder::build(element));
+            if (pSetting.get() == nullptr) {
+                qDebug() << "Ignoring unsupported controller setting in file"
+                         << pMapping->filePath() << "at line"
+                         << element.lineNumber() << ".";
+                continue;
+            }
+            if (!pSetting->valid()) {
+                qDebug() << "The parsed setting in file" << pMapping->filePath()
+                         << "at line" << element.lineNumber()
+                         << "appears to be invalid. It will be ignored.";
+                continue;
+            }
+            if (pMapping->addSetting(pSetting)) {
+                layout->addItem(pSetting);
+            } else {
+                qDebug() << "The parsed setting in file" << pMapping->filePath()
+                         << "at line" << element.lineNumber()
+                         << "couldn't be added. Its layout information will also be ignored.";
+                continue;
+            }
+        } else if (tagName == "row") {
+            LegacyControllerSettingsLayoutContainer::Disposition orientation =
+                    element.attribute("orientation").trimmed().toLower() ==
+                            "vertical"
+                    ? LegacyControllerSettingsLayoutContainer::VERTICAL
+                    : LegacyControllerSettingsLayoutContainer::HORIZONTAL;
```

Comment:
I don't think this is needed. Grouping abilities make sense, but rows/orientation should be controlled by Mixxx and not be the script developer's concern. If we expose such functionality, we will need to support this for a long time, even with QML.

Reply:
We discussed this above and the outcome was that we want no complex layout here, but vertical and horizontal grouping. 
@acolombier implemented what was the outcome of the discussion and also spend a lot of work addressing many detail remarks as it operation on any screen size etc.
Now we are in the final steps of the code review.

Reply:
Alright. I would have preferred it without such functionality, as you could just put it in a separate line, but you already discussed it. Just keep in mind that it may turn out difficult to implement this in QML.

Reply:
Yes, that makes sense. I imagine this is something that should be looked at while reviewing controller contributions and make sure settings remain simple enough and don't abuse this feature. Perhaps it might be worth also add this suggestion in the controller wiki, but I don't think I have permission to make this change.